scripts/uosc/elements/Menu.lua

```diff
@@ -608,6 +620,104 @@ function Menu:on_end()
 	if self.current.selected_index > 0 then self:scroll_to_index(self.current.selected_index) end
 end
 
+---@param menu MenuStack
+function Menu:search_internal(menu)
+	local query = menu.search.query
+	menu.items = query ~= '' and itable_filter(menu.search.source.items, function(item)
+		return item.title and item.title:lower():find(query, 1, true) or
+			item.hint and item.hint:lower():find(query, 1, true)
+	end) or menu.search.source.items
+	self:update_items(self.root.items)
+end
+
+---@param menu MenuStack
+function Menu:perform_search(menu)
+	menu = menu or self.current
+	if self.opts.on_search then
+		if self.opts.search_debounce ~= 0 then
+			menu.items = {
+				{
+					title = 'Loading...',
+					icon = 'spinner',
+					italic = true,
+					align = 'center',
+					selectable = false,
+					muted = true
+				}
+			}
+			self:update_items(self.root.items)
+		end
+		self.opts.on_search(menu.id, menu.search.query)
+	else self:search_internal(menu) end
+end
+
+---@param query string
+function Menu:search_query_update(query)
+	local menu = self.current
+	menu.search.query = query
+	if self.opts.search_debounce == 0 then
+		self:perform_search(menu)
+	elseif self.opts.search_debounce then
+		menu.search.timeout:kill()
+		menu.search.timeout:resume()
+	end
+	request_render()
+end
+
+function Menu:search_backspace()
+	local pos, search_text = #self.current.search.query - 1, self.current.search.query
+	while pos > 1 and search_text:byte(pos) >= 0x80 and search_text:byte(pos) <= 0xbf do
+		pos = pos - 1
+	end
+	self:search_query_update(search_text:sub(1, pos))
+end
+
+function Menu:search_ensure_keybinds()
+	if self.current.search then
+		self:add_key_binding('esc', 'menu-search-close', self:create_key_action('search_stop'))
+		self:add_key_binding('bs', 'menu-search-bs', self:create_key_action('search_backspace'), {repeatable = true})
+		self:add_key_binding('ctrl+enter', 'menu-search-submit', self:create_key_action('perform_search'))
+	else
+		self:search_disable_key_bindings()
+	end
+end
+
+function Menu:search_text_input(info)
+	if info.key_text and (info.event == 'press' or info.event == 'down' or info.event == 'repeat') then
+		local menu = self.current
+		if not menu.search then
+			local timeout
+			if self.opts.search_debounce and self.opts.search_debounce > 0 then
+				timeout = mp.add_timeout(self.opts.search_debounce / 1000, self:create_action(function()
+					self:perform_search(menu)
+				end))
+			end
+			menu.search = {
+				query = '', timeout = timeout, source = {selected_index = menu.selected_index, items = menu.items},
+			}
+			self:search_ensure_keybinds()
+		end
+		self:search_query_update(menu.search.query .. info.key_text)
+	end
+end
+
+function Menu:search_stop()
+	self:search_query_update('')
+	self.current.search = nil
+	self:search_ensure_keybinds()
+end
+
+function Menu:search_disable_key_bindings()
```

Comment:
I don't like the keybind juggling. It's fickle and leads to unstable states. Like searching menu, than going to submenu, searching that, than back to menu, canceling search there, then back to submenu, now when I try to cancel search here we instead exit the menu.

Rather, we bind all the bindings we need at init as we do now, and just fork the search specific handling in their handles if search exists.

And `ctrl+enter` is currently unused even when it's defined. It just falls back to normal open item handling, ignoring the ctrl modifier.

Reply:
The scenario you describe with canceling search is caused by that one issue I described, otherwise going back to the submenu and canceling search there would work as expected. That's why `search_ensure_keybinds()` exists.

And what do you mean that `ctrl+enter` ignores ctrl? I've tested the 'submit' search and it works as it should.